### ISSUE TITLE
[Ray Client] Using many concurrent client connections results in deadlock/hanging

### ISSUE THREAD

**Issue by kiralaz** [ray-project/ray#26144]
### What happened + What you expected to happen

This is not a contribution.

We have some internal use cases for communication across ray clusters and recently ran into an issue where ray client connections inexplicably deadlock/hang indefinitely. 

We noticed that when we make a bunch of client connections around the same time, we consistently get this hanging issue. In the reproduction script below we make 50 concurrent connections, but internally we were planning to have use cases relying on this functionality that would potentially make much more than that. In some manual testing we noticed that lower numbers (up to 25-30) worked consistently, while 40+ would sometimes hang and 50 consistently hangs.

Note that this occurs in all of the following cases

* Making client connections from actors on the same cluster back to its own head node (not our use case, but useful for reproduction)
* Making client connections originating from actors on a different cluster
* A mix of the two

### Versions / Dependencies

ray==1.12.1, ray==1.13.0

Python 3.9.4

Ubuntu 18.04

### Reproduction script

The following script can reproduce the issue on a single machine. Similar setups in multi-nodes or cluster(s) should reproduce the issue as well, and was how we originally came across the issue. 
First, on the machine run
```
ray start --head
```
Then `python3 ray_client_hanging.py`
where `ray_client_hanging.py` contains
```
# ray_client_hanging.py
import ray

HEAD_NODE = 'localhost'


class ClientMakingActor:
    def run(self):
        client = ray.init(f'ray://{HEAD_NODE}:10001', allow_multiple=True)
        client.disconnect()


def repro():
    ray.init(address="auto")
    actors = []
    for _ in range(50):
        actors.append(ray.remote(ClientMakingActor).remote())

    refs = [a.run.remote() for a in actors]
    print(ray.get(refs))


if __name__ == "__main__":
    # Run ray start --head first
    repro()
```

this should work in the multi-node cluster setting as well, if you replace `HEAD_NODE` with hostname or IP of the head node.

Using `py-spy dump --pid $PID` where `$PID` is the pid of any ClientMakingActor process, this is an example stack trace of the hanging client connection.
```
Process 3438: ray::ClientMakingActor.run()
Python v3.9.4 (/usr/bin/python3.9)

Thread 0x7FCB8C98F740 (active): "MainThread"
    _blocking (grpc/_channel.py:910)
    __call__ (grpc/_channel.py:921)
    get_cluster_info (ray/util/client/worker.py:702)
    is_initialized (ray/util/client/worker.py:748)
    is_initialized (ray/util/client/api.py:202)
    wrapper (ray/_private/client_mode_hook.py:104)
    client_mode_should_convert (ray/_private/client_mode_hook.py:122)
    _remote (ray/remote_function.py:269)
    _invocation_remote_span (ray/util/tracing/tracing_helper.py:303)
    remote (ray/remote_function.py:227)
    connect (ray/client_builder.py:167)
    init (ray/worker.py:882)
    wrapper (ray/_private/client_mode_hook.py:105)
    run (ray_client_hanging.py:8)
    _resume_span (ray/util/tracing/tracing_helper.py:462)
    actor_method_executor (ray/_private/function_manager.py:701)
    main_loop (ray/worker.py:449)
    <module> (ray/workers/default_worker.py:235)
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by wuisawesome** [comment]
I see, would it be possible to connect to all the clusters from the same driver process (like in the [docs](https://docs.ray.io/en/latest/cluster/ray-client.html#connect-to-multiple-ray-clusters-experimental))?

**Comment by kiralaz** [comment]
Yes, that's our primary use case, but we assumed it was also possible to create ray client connections across clusters to better support some other functionalities we had in mind. Is that not how ray client is intended to be used? Are there known limitations?

**Comment by scv119** [comment]
escalate it to p0 in case we forget about this one.

**Comment by iycheng** [comment]
I'll take a look.

**Comment by Hoeze** [comment]
Is this still an issue?

**Comment by hora-anyscale** [comment]
Per Triage Sync: @ckw017 can you please review this?

**Comment by ckw017** [comment]
> Per Triage Sync: @ckw017 can you please review this?

Think we need to just merge this https://github.com/ray-project/ray/pull/27940, but was blocked on committer approval, will try reopening.

**Comment by trent** [comment]
I am seeing this same behavior. I tried pulling the patch in https://github.com/ray-project/ray/pull/27940 and it did not fix the issue.

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
